rfc: api: define max pooling with global indices

dzarukin opened the pull request:
# Description

This RFC proposes a way to support global indices feature for a max pooling operation.

Link to a rendered document: https://github.com/oneapi-src/oneDNN/tree/dzarukin/rfcs/max_pooling_global_indices/rfcs/20201011-max-pooling-with-global-indices

emfomenk commented on rfcs/20201011-max-pooling-with-global-indices/README.md:
```diff
@@ -0,0 +1,66 @@
+### Option 2 - re-use existing algorithm
+
+This option makes `dnnl_pooling_max` algorithm to dump global indices
+unconditionally if the second output memory is not empty. The drawback is
```
we don't have "conditional" inputs or outputs. each pd() fully defines the expected inputs and outputs (we also have a check in `cvt_primitive_args()` function in `src/common/primitive_exec_types.cpp`). I would recommend keeping it that way.

dzarukin replied:
Makes sense, rephrased the sentence that it will require from users to always specify two arguments with breaks API in some sense. Thanks for the comment.

emfomenk commented on rfcs/20201011-max-pooling-with-global-indices/README.md:
```diff
@@ -0,0 +1,66 @@
+} dnnl_alg_kind_t;
+~~~
+It will instruct the implementation to dump global indices in a second output
+marked as `DNNL_ARG_DST_1`. By introducing a new algorithm we preserve
```
Do we know what is the order of the indices should be? Say, we typically use NCHW "natural" order. But what if TF uses NHWC -- would that affect the way a particular index is translated to a number. In other words, could it be that the same conceptual input will produce different "global indices" depending on the framework?

dzarukin replied:
Once you asked, now I've understood it's a bit of a problem.

What if we don't do global indices in sense of absolute numbers staying one by one reflecting kernel position but instead return the memory marking the physical position of the biggest value with `1`. This way users can reorder memory the way they need. It will be some sort of what workspace memory now stash but with ability to specify the data type and physical format where the actual global number is a physical offset of the point. When keeping the destination memory and global indices memory in the same memory format, the access can be done via a single offset.

Alternative approach would be to store global dim_t values instead but use memory descriptor format as a way for enumeration or to specify permuted logical axes values to be able to to reflect the physical format of destination memory.

emfomenk replied:
> What if we don't do global indices in sense of absolute numbers staying one by one reflecting kernel position but instead return the memory marking the physical position of the biggest value with 1. 

That's an option and interesting idea. But, I am afraid it's too complicated and the corresponding translation will kill all the performance benefits of using oneDNN (I am actually afraid that producing the "proper" global indices will kill the performance benefits anyways).

I think we should first check what the main frameworks do and see if there is a pattern. If there would be too many cases to support, we could consider the alternatives, or just admit that this flavor of max pooling is not worth implementing in oneDNN at all.

> Alternative approach would be to store global dim_t values instead but use memory descriptor format as a way for enumeration or to specify permuted logical axes values to be able to to reflect the physical format of destination memory.

Not sure if this addresses the issue: converting the data in to another layout != converting the array of the indices, but changing the number there (if I understand things correctly). Example:

```
// max pooling: 2x2
A = [0 1]
    [3 2]   // row-major order
global max = [2] ("3" is at position (1,0) --> 2)

but the same array in col-major will give:
global max = [1] // cannot be obtain from "reordering" the global max array above
```

dzarukin replied:
I've updated an RFC with Implementation section after our discussion, putting my best English to describe the problem and drawbacks.

emfomenk commented on rfcs/20201011-max-pooling-with-global-indices/README.md:
```diff
@@ -0,0 +1,66 @@
+whether the memory should be read or written in `execute()` (e.g. `with_bias()`
+method for convolutions), so run-time check is less intuitive from API
+perspective at the same time.
+
```
Probably worth mentioning the 3rd option (though I think option 1 is still preferable).

We could add an auxiliary primitive that would translate workspace from the pooling to a memory with global indices. 

Internally I am almost 99% sure that would be how we will support this feature. The question is: do we want to hide this "extra" primitive or we just expose it.

dzarukin replied:
It's a good point. I've added a third option, thanks.

dzarukin commented on rfcs/20201011-max-pooling-with-global-indices/README.md:
```diff
@@ -0,0 +1,66 @@
+## Motivation
+
+There are some
+[neural networks for semantic segmentation](Learning_Deconvolution_Network_for_Semantic_Segmentation.pdf)
```
Thanks for the link and concern. Updated.

georgen117 commented:
The only part of the RFC I am a little confused about is the `dnnl_pooling_inted_value_t`

```
typedef enum {
    /// Pooling to use the full tensor logical offset
    dnnl_pooling_index_value_all_dims,
    /// Pooling to use the tensor logical offset without mini-batch (dims[0])
    dnnl_pooling_index_value_c_spatial,
    /// Pooling to use the tensor logical offset of spatial values (dims[2 - N])
    dnnl_pooling_index_value_spatial,
} dnnl_pooling_index_value_t;
```

Please help make sure I am clear about what each options means.
 - `dnnl_pooling_index_value_all_dims` - If I am reading this right this matches up with ONNX definition and one of the TensorFlow definitions of Indices. It uses all dimensions of the tensor when populating a particular indice. (i.e. for nchw the indices would we [0, N x C x H x W) )
- `dnnl_pooling_index_value_c_spatial` - This matches the second TensorFlow definition. (i.e. for nchw the indices would be [0, C x H x W) )
- `dnnl_pooling_index_value_spatial` - This matches the Pytorch definition (i.e. for nchw the indices woudl be [0, H x W ) )

Am I reading the `dnnl_poolingindex_value_t` correct?

dzarukin commented:
Hi @georgen117, yes, your comment explains the thing exactly as it's envisioned. Do you think there could be better names to describe them more clear? Thanks.

georgen117 commented:
@dzarukin off the top of my head I don't have a suggestion for better names.  I think adding to the description may help. Something along the lines of the i.e. sections I added in my question above.  Should the name also be changed to `dnnl_pooling_indices_value_xxx` (index --> indices)?  The enum is referring to the output indices right.

I am still quite new to ML so index and indices may be used interchangeable for pooling and I am just not familiar with that fact yet.

dzarukin commented:
> @dzarukin off the top of my head I don't have a suggestion for better names. I think adding to the description may help. Something along the lines of the i.e. sections I added in my question above. Should the name also be changed to `dnnl_pooling_indices_value_xxx` (index --> indices)? The enum is referring to the output indices right.
> 
> I am still quite new to ML so index and indices may be used interchangeable for pooling and I am just not familiar with that fact yet.

I've added a bit more documentation notes that show what is the available range of values for each of options. Thanks for the input.

igorsafo commented on rfcs/20201011-max-pooling-with-global-indices/README.md:
```diff
@@ -0,0 +1,274 @@
+
+## Open Question
+
+* Are getter functions required for attributes extension?
```
I would vote for adding getter functions for API completeness and simplifying interface tests of set functions. Also it is pretty straight forward from implementation point of view.

dzarukin replied:
Alright, updated an RFC with getter counterparts, thanks.

igorsafo reviewed (approved):
RFC looks good to me.

Did we get a feedback from customers (ONNX, TF, PyTorch, OpenVino)?

dzarukin commented:
We received the good-to-go feedback, though customers wished we had performance numbers (which would obviously require some prototype we haven't coded yet).
